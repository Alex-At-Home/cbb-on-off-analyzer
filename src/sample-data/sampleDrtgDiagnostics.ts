export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
