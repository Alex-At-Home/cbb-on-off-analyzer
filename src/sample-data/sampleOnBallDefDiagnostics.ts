/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOnBallDefDiagnostics =
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
};
