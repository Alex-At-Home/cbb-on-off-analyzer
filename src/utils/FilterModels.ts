import { DateUtils } from "./DateUtils";

/** Typescript limitations - also have to repeat this for ParamPrefixesType */
export class ParamPrefixes {
  static readonly game = "game-"; //(this is on-off / team info)
  static readonly lineup = "lineup-";
  static readonly report = "report-"; //(not used currrently, we re-use lineup)
  static readonly roster = "roster-";
  static readonly player = "player-";
}
export type ParamPrefixesType = "game-" | "lineup-" | "report-" | "roster-" | "player-";

/** The common luck config */
export type LuckParams = {
  base: "baseline" | "season"
};

export type ManualOverride = {
  rowId: string, //(the player key, lineup key, or on/off key)
  statName: string, //(the field that has been changed)
  newVal: number,
  use: boolean
}

/** Common params across all filter types */
export type CommonFilterParams = {
  year?: string,
  team?: string,
  gender?: string,
  minRank?: string,
  maxRank?: string,
  baseQuery?: string,
  filterGarbage?: boolean, //(missing iff "false")
  queryFilters?: string, //(missing iff empty)

  // FOR INTERNAL USE ONLY
  invertBase?: string //special case - the inverse of these two are used in combination with the usual queryFilters/baseQuery
  invertBaseQueryFilters?: string
};

/** Extracts the common params from a superset */
export function getCommonFilterParams(p: CommonFilterParams) {
  return {
    year: p.year,
    team: p.team,
    gender: p.gender,
    minRank: p.minRank,
    maxRank: p.maxRank,
    baseQuery: p.baseQuery,
    filterGarbage: p.filterGarbage,
    queryFilters: p.queryFilters
  };
}

/** Extracts the common leaderboard params from a superset */
export function getCommonLboardFilterParams(p: CommonFilterParams, tier?: string) {
  return {
    tier: tier,
    year: p.year,
    gender: p.gender,
  };
}

/** Combined params for game filtering */
export type GameFilterParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  onQuery?: string,
  offQuery?: string,
  onQueryFilters?: string, //(missing iff empty)
  offQueryFilters?: string, //(missing iff empty)
  autoOffQuery?: boolean
  // Team view
  teamDiffs?: boolean,
  showExtraInfo?: boolean,
  showTeamPlayTypes?: boolean,
  showRoster?: boolean,
  showGrades?: string,
  // Manual override:
  manual?: ManualOverride[],
  showPlayerManual?: boolean,
  showOnBallConfig?: boolean,
  // Global luck adjustments
  luck?: LuckParams, //(missing iff default)
  // Luck adjustments
  onOffLuck?: boolean,
  showOnOffLuckDiags?: boolean,
  showPlayerOnOffLuckDiags?: boolean,
  showPlayerPlayTypes?: boolean,
  // Misc display:
  showInfoSubHeader?: boolean,
  // Filtering of individual view:
  filter?: string,
  sortBy?: string,
  showBase?: boolean,
  showExpanded?: boolean,
  showDiag?: boolean
  possAsPct?: boolean,
  factorMins?: boolean,
  showPosDiag?: boolean,
  calcRapm?: boolean,
  // For leaderboard building:
  getGames?: boolean
};

/** Params for lineup filtering */
export type LineupFilterParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  // These params need to be explicitly merged in buildParamsFromState(true)
  // For sorting in the generated table:
  decorate?: boolean,
  showTotal?: boolean,
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string,
  // Filtering:
  filter?: string,
  // Luck adjustments
  luck?: LuckParams, //(missing iff default)
  lineupLuck?: boolean,
  showLineupLuckDiags?: boolean,
  // Other features:
  aggByPos?: string,
  showGameInfo?: boolean
};

export type LineupLeaderboardParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  tier?: string, //High, Medium, Low
  conf?: string, //(undefined ==> all conferences)
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string,
  // Filtering:
  filter?: string,
  lineupFilters?: string,
  // Luck adjustments
  showLineupLuckDiags?: boolean,
  // Query pre-sets
  confOnly?: boolean,
  t100?: boolean
};

export type TeamEditorParams = {
  [P in keyof PlayerLeaderboardParams]?: PlayerLeaderboardParams[P];
} & {
  // Player editor settings:
  showPrevSeasons?: boolean, //(defaults to false)
  offSeason?: boolean, //(defaults to true, else shows current performance - for building all star teams and seeing effect of injury)
  evalMode?: boolean, //(defaults to false, if true will compare the season X with the predictions from the previous offseason)
  alwaysShowBench?: boolean, //(defaults to false)
  superSeniorsBack?: boolean, //(defaults to false)
  // Controls what transfers are shown
  showOnlyTransfers?: boolean, //(defaults to true)
  showOnlyCurrentYear?: boolean, //(defaults to true)
  diffBasis?: string, //JSON representation of the starting point for showing diffs
  // Editor state
  deletedPlayers?: string, //;-separated list
  disabledPlayers?: string,//;-separated list
  addedPlayers?: string, //;-separated list
  editOpen?: string, //;-separated list, <key>|<open-tab>
  overrides?: string; //;-separated list, see TeamEditorUtils.PlayerEditModel
  allEditOpen?: string, //(defaults to undefined, all edit pages open to the tab if true)
};

export type OffseasonLeaderboardParams = {
  year?: string,
  teamView?: string,
  confs?: string,
  evalMode?: boolean,
} & Record<string, string>;

export type PlayerLeaderboardParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  tier?: string,  //All, High, Medium, Low
  conf?: string, //(undefined ==> all conferences)
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string,
  // Player settings
  posClasses?: string, //(undefined => all positions)
  possAsPct?: boolean,
  factorMins?: boolean,
  useRapm?: boolean,
  // Filtering:
  filter?: string,
  advancedFilter?: string,
  // Misc display:
  showInfoSubHeader?: boolean,
  // Query pre-sets
  confOnly?: boolean,
  t100?: boolean,
  // Transfer info
  transferMode?: string //==true => show only available, vs ==$year show all
};

export type TeamLeaderboardParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  conf?: string, //(undefined ==> all conferences)
  // Lots of settings:
  qualityWeight?: string,
  pinQualityWeight?: string,
  wabWeight?: string,
  pinWabWeight?: string,
  waeWeight?: string,
  pinWaeWeight?: string,
  domWeight?: string,
  pinDomWeight?: string,
  timeWeight?: string
  pinTimeWeight?: string
};


export type TeamReportFilterParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  // These params need to be explicitly merged in buildParamsFromState(true)
  // For sorting in the generated table:
  sortBy?: string,
  // Filtering:
  filter?: string,
  showOnOff?: boolean,
  showComps?: boolean,
  incRepOnOff?: boolean,
  regressDiffs?: string, //+ve to add that number of 0 samples, -ve to regress to the given sample size
  repOnOffDiagMode?: string //(the number of diagnostic lineups to show, basically 0 or 20:sort order:sort field)
  incRapm?: boolean,
  rapmDiagMode?: string, //"" if disabled, "team" if enabled with nobody expanded, "playerId[;playerId]+" if expanded for players
  rapmPriorMode?: string, //(-1==default==adapative, else the prior weight as 0->1)
  // Luck adjustments
  luck?: LuckParams, //(missing iff default)
  teamLuck?: boolean
  //(there's no luck diags here because we're applying at the lineup level)
};

/** Used to give compile errors if a field is omitted, for fw compat */
export type RequiredTeamReportFilterParams = {
  [P in keyof TeamReportFilterParams]?: TeamReportFilterParams[P];
} & {
  [P in keyof Required<CommonFilterParams>]: CommonFilterParams[P] | undefined;
};

export class ParamDefaults {
  // Game
  static readonly defaultAutoOffQuery = true;
  static readonly defaultPlayerFilter = "";
  static readonly defaultPlayerSortBy = "desc:off_team_poss_pct:baseline";
  static readonly defaultOnOffLuckAdjust = false;
  static readonly defaultOnOffLuckDiagMode = false;
  static readonly defaultTeamShowPlayTypes = false;
  static readonly defaultTeamShowRoster = false;
  static readonly defaultTeamEnabledGrade = "rank:Combo";
  static readonly defaultPlayerShowPlayTypes = false;
  static readonly defaultPlayerDiagMode = false;
  static readonly defaultPlayerShowBase = false;
  static readonly defaultPlayerShowExpanded = false;
  static readonly defaultPlayerPossAsPct = true;
  static readonly defaultPlayerFactorMins = false;
  static readonly defaultPlayerPosDiagMode = false;
  static readonly defaultPlayerCalcRapm = false;
  // Lineup
  static readonly defaultLineupShowTotal = false;
  static readonly defaultLineupDecorate = true;
  static readonly defaultLineupMinPos = "5";
  static readonly defaultLineupMaxTableSize = "50";
  static readonly defaultLineupSortBy = "desc:off_poss";
  static readonly defaultLineupFilter = "";
  static readonly defaultLineupLuckAdjust = false;
  static readonly defaultLineupLuckDiagMode = false;
  static readonly defaultLineupAggByPos = "";
  static readonly defaultLineupShowGameInfo = false;
  // leaderboards
  static readonly defaultTier = "High";
  // Lineup leaderboard
  static readonly defaultLineupLboardMinPos = "20";
  static readonly defaultLineupLboardMaxTableSize = "100";
  static readonly defaultLineupLboardSortBy = "desc:diff_adj_ppp";
  static readonly defaultLineupLboardFilter = "";
  static readonly defaultLineupLboardLuckDiagMode = false;
  // Player leaderboard
  static readonly defaultPlayerLboardMinPos = "20";
  static readonly defaultPlayerLboardMaxTableSize = "100";
  static defaultPlayerLboardSortBy(useRapm: boolean, factorMins: boolean) {
    return useRapm ? (factorMins ? "desc:diff_adj_rapm_prod" : "desc:diff_adj_rapm") : (factorMins ? "desc:off_adj_prod" : "desc:off_adj_rtg");
  }  
  static readonly defaultPlayerLboardFilter = "";
  static readonly defaultPlayerLboardFactorMins = false;
  static readonly defaultPlayerLboardPossAsPct = true;
  static readonly defaultPlayerLboardUseRapm = true;
  // Team leaderboard
  static readonly defaultTeamLboardQualityWeight = "0.5";
  static readonly defaultTeamLboardDomWeight = "0.25";
  static readonly defaultTeamLboardWabWeight = "0.5"; //(don't have decimal places for comparison with "" + 1.0)
  static readonly defaultTeamLboardWaeWeight = "0.25";
  static readonly defaultTeamLboardTimeWeight = "0"; //(don't have decimal places for comparison with "" + 0.0)
  // Report
  static readonly defaultTeamReportSortBy = "desc:off_poss:on";
  static readonly defaultTeamReportFilter = "";
  static readonly defaultShowOnOff = true;
  static readonly defaultShowComps = false;
  static readonly defaultTeamReportIncRepOnOff = false;
  static readonly defaultTeamReportIncRapm = false;
  static readonly defaultTeamReportRegressDiffs = "-2000";
  static readonly defaultTeamReportRepOnOffDiagMode = "0";
  static readonly defaultTeamReportRepOnOffDiagModeIfEnabled = [ "20", "-1", "lineup.off_poss.value" ];
  static readonly defaultTeamReportRapmDiagMode = "";
  static readonly defaultTeamReportRapmPriorMode = "-1";
  static readonly defaultTeamReportLuckAdjust = false;
  // Common
  static readonly defaultTeam = "";
  static readonly defaultYear = DateUtils.mostRecentYearWithData;
  static readonly defaultLeaderboardYear = DateUtils.mostRecentYearWithLboardData; //(takes a while longer to get updated)
  static readonly defaultGender = "Men";
  static readonly defaultMinRank = "0";
  static readonly defaultMaxRank = "400";
  static readonly defaultFilterGarbage = false;
  static readonly defaultQueryFilters = "";
  static readonly defaultLuckConfig: LuckParams = { base: "season" };
};

export type FilterParamsType = GameFilterParams | LineupFilterParams | TeamReportFilterParams;

/** Which API to call and with what object */
export type FilterRequestInfo = {
  context: ParamPrefixesType,
  paramsObj: FilterParamsType,
  includeRoster?: boolean //(if true will fetch the roster, eg one call per page should do this)
};
