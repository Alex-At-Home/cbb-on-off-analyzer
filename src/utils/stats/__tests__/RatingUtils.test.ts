

import _ from 'lodash';

import { RatingUtils } from "../RatingUtils";
import { GameFilterParams, LineupFilterParams, TeamReportFilterParams } from "../../FilterModels";
import { samplePlayerStatsResponse } from "../../../sample-data/samplePlayerStatsResponse";
import { sampleOrtgDiagnostics } from "../../../sample-data/sampleOrtgDiagnostics";
import { sampleDrtgDiagnostics } from "../../../sample-data/sampleDrtgDiagnostics";

describe("RatingUtils", () => {

  // Write the data objects out in pure JS format so we can store them
  expect.addSnapshotSerializer({
    test: (val: any) => true,
    print: (val: any) => JSON.stringify(val, null, 3)
  });

  test("RatingUtils - buildORtg", () => {
    const playerInfo = _.cloneDeep(
      samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets[0]
    );
    const [ oRtg, adjORtg, rawORtg, rawAdjORtg, oRtgDiags ] = RatingUtils.buildORtg(
      playerInfo, 100, true, false
    );
    const expORtg = {value:113.40865079767507};
    const expORtgAdj = {value:5.002326978989183};
    expect(oRtg).toEqual(expORtg);
    expect(adjORtg).toEqual(expORtgAdj);
    expect(rawORtg).toEqual(undefined);
    expect(rawAdjORtg).toEqual(undefined);
    expect(oRtgDiags).toMatchSnapshot();
    // Then can be copied into here:
    expect(oRtgDiags).toEqual(sampleOrtgDiagnostics);

    // Check with override:
    (playerInfo as any).off_3p = {
      value: playerInfo.off_3p.value - 0.1,
      old_value: playerInfo.off_3p.value
    };
    const [ oRtg2, adjORtg2, rawORtg2, rawAdjORtg2, oRtgDiags2 ] = RatingUtils.buildORtg(
      playerInfo, 100, true, true
    );
    expect(oRtg2).toEqual({value:104.9838324244601});
    expect(adjORtg2).toEqual({value:3.123247651476328});
    expect(rawORtg2).toEqual(expORtg);
    expect(rawAdjORtg2).toEqual(expORtgAdj);
  });
  test("RatingUtils - buildOffOverrides", () => {
    const outputs = {
      total_off_fgm: { value: 0.0001 },
      total_off_2p_made: { value: 0.0002 },
      total_off_3p_made: { value: 0.0003 },
      total_off_ftm: { value: 0.0004 },
      team_total_off_pts: { value: 0.0005 },
      team_total_off_fgm: undefined, //(check it will be set as 0)
      team_total_off_3p_made: { value: 0.0007 },
      team_total_off_ftm: { value: 0.0008 },
      team_total_off_to: { value: 0.0009 }
    };
    const testStatSet = {
      // some volume numbers:
      total_off_3p_attempts: { value: 10 },
      total_off_2p_attempts: { value: 20 },
      total_off_fta: { value: 20 },
      total_off_to: { value: 20 },
      off_poss: { value: 100 },

      // all the possible overrides currently:
      off_3p: { value: 0.5, old_value: 0.4 },
      off_2p: { value: 0.6, old_value: 0.4 },
      off_ft: { value: 0.9, old_value: 0.7 },
      off_to: { value: 0.25, old_value: 0.2 },

      // Outputs:
      ...outputs
    };
    expect(RatingUtils.buildOffOverrides(testStatSet)).toEqual({
      total_off_to: { value: 26.666666666666668 },
      off_poss: { value: 106.66666666666667 }, //(extra 5% TOs)

      total_off_fgm: { value: 5.000099999999999 },
      total_off_2p_made: { value: 4.0001999999999995 },
      total_off_3p_made: { value: 1.0002999999999997 },
      total_off_ftm: { value: 4.000400000000002 },
      team_total_off_pts: { value: 15.000499999999999 },
      team_total_off_fgm: { value: 4.999999999999999 },
      team_total_off_3p_made: { value: 1.0006999999999997 },
      team_total_off_ftm: { value: 4.000800000000002 },
      team_total_off_to: { value: 6.667566666666667 }
    });
    // Now just check it does nothing if there are no changes
    const testStatSet2 = {
      ...testStatSet,
      off_3p: { value: 0.4 },
      off_2p: { value: 0.4 },
      off_ft: { value: 0.7 },
      off_to: { value: 0.2 }
    };
    expect(RatingUtils.buildOffOverrides(testStatSet2)).toEqual({
      ...outputs,
      team_total_off_fgm: { value: 0 },
      total_off_to: { value: 20 },
      off_poss: { value: 100 },
    });
  });

  test("RatingUtils - buildDRtg", () => {
    const playerInfo = _.cloneDeep(
      samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets[0]
    );
    const [ dRtg, adjDRtg, rawDRtg, rawAdjDRtg, dRtgDiags ] = RatingUtils.buildDRtg(
      playerInfo, 100, true, false
    );
    const expDRtg = {value:98.84955759282161};
    const expDRtgAdj = {value:-1.7012563565592531};
    expect(dRtg).toEqual(expDRtg);
    expect(adjDRtg).toEqual(expDRtgAdj);
    expect(rawDRtg).toEqual(undefined);
    expect(rawAdjDRtg).toEqual(undefined);
    expect(dRtgDiags).toMatchSnapshot();
    // Then can be copied into here:
    expect(dRtgDiags).toEqual(sampleDrtgDiagnostics);

    // Check with override:
    (playerInfo as any).oppo_def_3p = {
      value: 0.3,
      old_value: 0.4
    };
    const [ dRtg2, adjDRtg2, rawDRtg2, rawAdjDRtg2, dRtgDiags2 ] = RatingUtils.buildDRtg(
      playerInfo, 100, true, true
    );
    expect(dRtg2).toEqual({value:89.6487161188261});
    expect(adjDRtg2).toEqual({value:-3.404489466919759});
    expect(rawDRtg2).toEqual(expDRtg);
    expect(rawAdjDRtg2).toEqual(expDRtgAdj);
  });
});
