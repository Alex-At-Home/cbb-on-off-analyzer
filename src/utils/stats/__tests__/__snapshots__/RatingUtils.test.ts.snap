// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingUtils RatingUtils - buildDRtg 1`] = `
{
   "stl": 30,
   "blk": 5,
   "drb": 96,
   "pfPct": 0.15789473684210525,
   "playerRtg": 105.86789969595162,
   "playerDelta": 1.7545855257825025,
   "scPossConceded": 0.48933399202505,
   "noShotCredit": 32.260450315336996,
   "reboundCredit": 34.59731465653162,
   "missFtCredit": 2.5382812499999994,
   "stopsIndPct": 0.19384370453035926,
   "stopsTeamPct": 0.3168223034445907,
   "teamBlk": 108,
   "oppoPts": 1738,
   "oppoPoss": 1790,
   "oppoFga": 1588,
   "oppoFgm": 637,
   "oppoFtm": 270,
   "oppoFta": 384,
   "oppoFtPoss": 182.39999999999998,
   "oppoTov": 312,
   "teamStl": 126,
   "teamDrb": 726,
   "opponentOrbPct": 0.274,
   "opponentFgPct": 0.40113350125944586,
   "teamOrbCreditToDefender": 0.29122292191435767,
   "teamOrbCreditToRebounder": 0.16408942065491183,
   "teamDvsRebCredit": 0.6396113056611289,
   "oppoFgMiss": 843,
   "oppoNonStlTov": 186,
   "teamMissWeight": 0.45209006306739913,
   "oppoFtPct": 0.703125,
   "oppoFtHitOnePlus": 166.32421874999997,
   "oppoProbFtHitOnePlus": 0.911865234375,
   "oppoScPoss": 803.32421875,
   "oppoPtsPerScore": 2.163510024264409,
   "teamRtg": 97.09497206703911,
   "dRtg": 98.84955759282161,
   "offSos": 108.03972067039106,
   "avgEff": 100,
   "adjDRtg": 91.49371821720374,
   "adjDRtgPlus": -1.7012563565592531
}
`;

exports[`RatingUtils RatingUtils - buildORtg 1`] = `
{
   "rawFga": 356,
   "rawFgx": 217,
   "rawFgm": 139,
   "ptsFgm": 334,
   "rawFtm": 172,
   "rawAssist": 147,
   "rawAssistInfo": [
      "70",
      "10",
      "67"
   ],
   "rawPts": 506,
   "rawOrb": 17,
   "rawTo": 69,
   "raw3Fga": 174,
   "raw2midFga": 65,
   "raw2rimFga": 117,
   "raw3Fgm": 56,
   "raw2midFgm": 18,
   "raw2rimFgm": 65,
   "teamOrb": 310,
   "teamPts": 1937,
   "teamFga": 1550,
   "teamFgm": 643,
   "teamFta": 579,
   "teamFtPct": 0.7564766839378239,
   "teamOrbPct": 0.3150406504065041,
   "teamTo": 299,
   "teamPoss": 1802,
   "eFG": 0.4691011235955056,
   "teamPtsPerScore": 2.175419074119295,
   "teamFtHitOnePlus": 247.4031517624634,
   "teamProbFtHitOnePlus": 0.9406963945340815,
   "rosterOrb": 0,
   "teamOrbCreditToRebounder": 0.2887736570583408,
   "teamOrbCreditToScorer": 0.18222190606216335,
   "teamScoreFromReboundPct": 0,
   "teamOrbWeight": 0.6131133277458456,
   "othersAssist": 214,
   "otherEfg": 0.48785594639866,
   "otherEfgInfo": [
      "48.8",
      "48.8",
      "48.8"
   ],
   "otherPtsPerFgm": 2.3115079365079363,
   "teamOrbContribPct": 0,
   "teamScoredPlayPct": 0.4215924014026815,
   "teamAssistRate_Classic": 0.3794090202177294,
   "ppFgTeamAstPct_Classic": 0.08899059884320337,
   "teamAssistRate": 0.37410071942446044,
   "ppFgTeamAstPct": 0.09475178477907331,
   "teamAssistedEfg": 0.5065576186265842,
   "ptsProd": 569.559794669086,
   "ppOrb": 9.55928445518132,
   "ppAssist": 88.78978224455611,
   "ppAssist_Classic": 82.88469256002232,
   "ppFg": 299.2107279693486,
   "ftPoss": 96.29706390328151,
   "actualFtaToPoss": 0.4542314335060449,
   "ftPct": 0.8113207547169812,
   "missedBothFTs": 0.03559985760056958,
   "offPlaysLessPoss": 69.81254009375766,
   "offPoss": 451.4845238095238,
   "fgPart": 125.8295019157088,
   "ftPart": 92.86890214097176,
   "astPart": 35.8574120603015,
   "astPart_Classic": 35.857412060301506,
   "orbPart": 4.394226642998126,
   "teamScoringPoss": 890.4031517624634,
   "teamPlays": 2112,
   "adjPoss": 475.2289159044038,
   "scoringPoss": 258.9500427599802,
   "fgxPoss": 143.85071138211381,
   "ftxPoss": 3.428161762309771,
   "oRtg": 119.84956630535876,
   "oRtg_Classic": 119.47176661703918,
   "defSos": 96.60184357541898,
   "avgEff": 100,
   "SD_at_Usage": 9.225388241385453,
   "SDs_Above_Mean": 2.1516239518585065,
   "SD_at_Usage_20": 10.143,
   "Regressed_ORtg": 121.82392174370084,
   "Usage": 26.37230387926769,
   "Usage_Bonus": 7.965379849084613,
   "adjORtg": 134.3548909513889,
   "adjORtgPlus": 6.870978190277782
}
`;

exports[`RatingUtils RatingUtils - buildOnBallDefenseAdjustmentsPhase1 1`] = `
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 0,
   "weightedUnadjDRtgMean": 0,
   "uncategorizedAdjustment": 0,
   "adjustedPossPct": 0,
   "dRtg": 0,
   "adjDRtg": 0,
   "adjDRtgPlus": 0
}
`;

exports[`RatingUtils RatingUtils - injectOnBallDefenseAdjustmentsPhase2 1`] = `
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
}
`;

exports[`RatingUtils RatingUtils - injectUncatOnBallDefenseStats 1`] = `
[
   {
      "code": "Code1",
      "title": "#1 Player1",
      "pts": 170,
      "plays": 216,
      "scorePct": 32.4,
      "tovPct": 15.7,
      "fgMiss": 111,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   },
   {
      "code": "Code2",
      "title": "#02 Player2",
      "pts": 159,
      "plays": 170,
      "scorePct": 41.2,
      "tovPct": 11.8,
      "fgMiss": 79,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   },
   {
      "code": "Code3",
      "title": "#13 Player3",
      "pts": 151,
      "plays": 169,
      "scorePct": 37.3,
      "tovPct": 16,
      "fgMiss": 78,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   },
   {
      "code": "Code4",
      "title": "#99 Player4",
      "pts": 109,
      "plays": 107,
      "scorePct": 39.3,
      "tovPct": 12.1,
      "fgMiss": 52,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   },
   {
      "code": "Code5",
      "title": "#21 Player5",
      "pts": 103,
      "plays": 148,
      "scorePct": 31.8,
      "tovPct": 14.9,
      "fgMiss": 79,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   },
   {
      "code": "Code6",
      "title": "#66 Player6",
      "pts": 52,
      "plays": 88,
      "scorePct": 26.1,
      "tovPct": 11.4,
      "fgMiss": 54,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   },
   {
      "code": "Code7",
      "title": "Player 7",
      "pts": 76,
      "plays": 95,
      "scorePct": 32.6,
      "tovPct": 14.7,
      "fgMiss": 50,
      "totalPts": 1554,
      "totalScorePct": 39.1,
      "totalPlays": 1780,
      "uncatPts": 734,
      "uncatPlays": 787,
      "uncatScorePct": 44.455654383735705,
      "uncatPtsPerScPlay": 2.097946070781385
   }
]
`;
